File: agent/src/provider/provider.py
from abc import ABC, abstractmethod
from typing import Optional
from src.grpc.agent_pb2 import (
    AgentApp,
    AgentInstance,
    AgentOS,
    AgentSSH,
    AgentWebItem,
)
from src.exec.exec import Exec


class Provider(ABC):
    def __init__(self, exec: "Exec"):
        self.exec = exec

    def get_instance(self) -> AgentInstance:
        return AgentInstance(
            os=self.get_os(),
            lan=self.get_lan(),
            web=self.get_web(),
            ssh=self.get_ssh(),
            apps=self.get_apps(),
            instances=self.get_instances(),
        )

    @abstractmethod
    def get_os(self) -> AgentOS:
        pass

    @abstractmethod
    def get_lan(self) -> str:
        pass

    @abstractmethod
    def get_web(self) -> list[AgentWebItem]:
        pass

    @abstractmethod
    def get_ssh(self) -> Optional[AgentSSH]:
        pass

    @abstractmethod
    def get_apps(self) -> list[AgentApp]:
        pass

    def get_instances(self) -> list[AgentInstance]:
        return []

File: agent/src/provider/provider.go
package provider

import (
	ex "agent/src/executor"
	gen "agent/src/grpc/generated"
)

type Provider interface {
	Check() bool
	GetInstance() gen.AgentInstance
	GetOS() gen.AgentOS
	GetLan() string
	GetWan() *string
	GetDDNS() *string
	GetDHCP() []*gen.AgentInstance_AgentDHCPItem
	GetWeb() []*gen.AgentWebItem
	GetSSH() *gen.AgentSSH
	GetApps() []*gen.AgentApp
	GetInstanceList() []*gen.AgentInstance
}

type BaseProvider struct {
	executor ex.Executor
}

func NewProvider(executor ex.Executor) *BaseProvider {
	return &BaseProvider{
		executor: executor,
	}
}

func (b *BaseProvider) GetInstance(e Provider) gen.AgentInstance {
	return gen.AgentInstance{
		Os:        e.GetOS(),
		Lan:       e.GetLan(),
		Wan:       e.GetWan(),
		Ddns:      e.GetDDNS(),
		Dhcp:      e.GetDHCP(),
		Web:       e.GetWeb(),
		Ssh:       e.GetSSH(),
		Apps:      e.GetApps(),
		Instances: e.GetInstanceList(),
	}
}
